src/range_object.cpp

```diff
@@ -3,11 +3,22 @@
 
 namespace Natalie {
 
-Value RangeObject::initialize(Env *env, Value begin, Value end, Value exclude_end_value) {
-    assert_not_frozen(env);
+RangeObject *RangeObject::create(Env *env, Value begin, Value end, bool exclude_end) {
+    assert_no_bad_value(env, begin, end);
+    auto range = new RangeObject(begin, end, exclude_end);
+    range->freeze();
+    return range;
+}
```

Comment:
I'm curious why we need a static function to create a RangeObject. Presumably RangeObject::initialize() would catch the invalid parameters, yes?

Reply:
Well we do not call `ìnitialize` when instantiating a Range from it's literal form (for example (1..2)). I thought of several solutions:

* Passing the env to the RangeObject constructor and call initialize in there. I do not really like passing the env into the constructor but it would work. I'm also not sure if throwing an exception in a constructor is fine but I guess it would be.
* Calling initialize after constructing the RangeObject at the caller site. This would lead to an allocation of the RangeObject even if it would be "invalid". Also we would probably forget about this when using RangeObject inside of other C++ code.
* Having a `create` function that takes the env, constructs the Range and calls initialize. This might also be a clean solution but we would also allocate the range before doing the bad value check.
* The current solution.

Which way would you prefer? Do you have other ideas?

Reply:
I like that you thought of all those alternatives. 🥰

I think the way you did it is fine, and your reasoning makes sense to me.

We may want to change the compiler so it emits code that creates objects with the equivalent of Klass.new in Ruby vs C++ `new Klass`. 🤔 

Let's ship this as is and give it all some thought down the road.